patches/server/1048-add-ItemFrameMeta-API.patch

```diff
@@ -0,0 +1,323 @@
++import java.io.ByteArrayInputStream;
++import java.io.ByteArrayOutputStream;
++import java.io.IOException;
++import java.util.Base64;
++import java.util.Map;
++import java.util.Objects;
++import java.util.Set;
++import java.util.logging.Level;
++import java.util.logging.Logger;
++import net.minecraft.core.component.DataComponentPatch;
++import net.minecraft.core.component.DataComponentType;
++import net.minecraft.core.component.DataComponents;
++import net.minecraft.nbt.CompoundTag;
++import net.minecraft.nbt.NbtAccounter;
++import net.minecraft.nbt.NbtIo;
++import net.minecraft.nbt.Tag;
++import net.minecraft.world.item.component.CustomData;
++import org.bukkit.Material;
++import org.bukkit.configuration.serialization.DelegateDeserialization;
++
++@DelegateDeserialization(SerializableMeta.class)
++public class CraftMetaItemFrame extends CraftMetaItem implements ItemFrameMeta {
++
++    private static final Set<Material> ENTITY_TAGGABLE_MATERIALS = Sets.newHashSet(
++        Material.ITEM_FRAME,
++        Material.GLOW_ITEM_FRAME
++    );
++
++    static final ItemMetaKeyType<CustomData> ENTITY_TAG = new ItemMetaKeyType<>(DataComponents.ENTITY_DATA, "EntityTag", "entity-tag");
++    CompoundTag entityTag;
++
++    static final ItemMetaKey ENTITY_ID = new ItemMetaKey("id", "entity-id");
++    static final ItemMetaKey FIXED = new ItemMetaKey("Fixed", "fixed");
++    static final ItemMetaKey INVISIBLE = new ItemMetaKey("Invisible", "invisible");
++
++    CraftMetaItemFrame(CraftMetaItem meta) {
++        super(meta);
++
++        if (meta instanceof CraftMetaItemFrame itemFrame) {
++            this.entityTag = itemFrame.entityTag;
++        }
++    }
++
++    CraftMetaItemFrame(DataComponentPatch tag, final Set<DataComponentType<?>> extraHandledDcts) {
++        super(tag, extraHandledDcts);
++
++        getOrEmpty(tag, CraftMetaItemFrame.ENTITY_TAG).ifPresent((nbt) -> this.entityTag = nbt.copyTag());
++    }
++
++    CraftMetaItemFrame(Map<String, Object> map) {
++        super(map);
++
++        String entityTag = SerializableMeta.getString(map, ENTITY_TAG.BUKKIT, true);
++        if (entityTag != null) {
++            ByteArrayInputStream buf = new ByteArrayInputStream(Base64.getDecoder().decode(entityTag));
++            try {
++                this.entityTag = NbtIo.readCompressed(buf, NbtAccounter.unlimitedHeap());
++            } catch (IOException ex) {
++                Logger.getLogger(CraftMetaItem.class.getName()).log(Level.SEVERE, null, ex);
++            }
++            return;
++        }
++        SerializableMeta.getObjectOptionally(Boolean.class, map, FIXED.BUKKIT, true).ifPresent((value) -> {
++            populateTagIfNull();
++            this.entityTag.putBoolean(FIXED.NBT, value);
++        });
++        SerializableMeta.getObjectOptionally(Boolean.class, map, INVISIBLE.BUKKIT, true).ifPresent((value) -> {
++            populateTagIfNull();
++            this.entityTag.putBoolean(INVISIBLE.NBT, value);
++        });
```

Comment:
None of this is needed. We are always serializing it as a full nbt tag instead of splitting it up. The only reason ArmorStand has similar logic is were weren't always doing that. But since this is brand new, we can do this.